### PR TITLE
Bug 1154088 - NSDate to relative time helper, last synced time periodic update

### PR DESCRIPTION
**Pull request by sachin004**
NSDate extension to convert nsdate to relative date and code to update last synced time in remote tabs panel.

### CODE REVIEW

**Comment by st3fan** [comment]
This looks good but before we use this I want to make sure that this functionality is not already provided by iOS through `NSDateFormatter.doesRelativeDateFormatting`.

**Comment by sachin004** [comment]
@st3fan by using <code>NSDateFormatter.doesRelativeDateFormatting</code> we can get only yesterday, today and tomorrow and remaining dates will be something like "12/2/2015 12:22:22".

**Comment by st3fan** [comment]
Ok thanks for the explanation @sachin004

File: Client/Frontend/Home/RemoteTabsPanel.swift [hunk 1]
```diff
@@ -63,6 +64,15 @@ class RemoteTabsPanel: UITableViewController, HomePanel {
         self.SELrefresh()
     }
 
+    override func viewDidAppear(animated: Bool) {
```
Comment (rnewman) [inline Client/Frontend/Home/RemoteTabsPanel.swift]: Can we split this into a separate bug? It has some potential side-effects, and I'd like to make sure we can land just the stringy parts.

File: Client/Frontend/Home/RemoteTabsPanel.swift [hunk 2]
```diff
@@ -63,6 +64,15 @@ class RemoteTabsPanel: UITableViewController, HomePanel {
...
+        timer = NSTimer.scheduledTimerWithTimeInterval(60.0, target: self.tableView, selector: Selector("reloadData"), userInfo: nil, repeats: true)
+    }
+
+    override func viewWillDisappear(animated: Bool) {
+        timer!.invalidate()
```
Comment (rnewman) [inline Client/Frontend/Home/RemoteTabsPanel.swift]: There's no harm in making this `timer?.invalidate()`.

File: ClientTests/RelativeDatesTests.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
```
Comment (rnewman) [inline ClientTests/RelativeDatesTests.swift]: Nit: Xcode screws up the leading spaces when you paste.

File: Utils/TimeConstants.swift [hunk 4]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
     public class func now() -> Timestamp {
         return UInt64(1000 * NSDate().timeIntervalSince1970)
     }
+    
+    public func toRelativeTimeString() -> String {
+        let now = NSDate()
+        let units = NSCalendarUnit.CalendarUnitSecond | NSCalendarUnit.CalendarUnitMinute | NSCalendarUnit.CalendarUnitDay | NSCalendarUnit.CalendarUnitWeekOfYear | NSCalendarUnit.CalendarUnitMonth | NSCalendarUnit.CalendarUnitYear | NSCalendarUnit.CalendarUnitHour
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: Nit: linebreaks!

File: Utils/TimeConstants.swift [hunk 5]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+        let components = NSCalendar.currentCalendar().components(units, fromDate: self, toDate: now, options: NSCalendarOptions.allZeros)
+        let formatter = NSDateFormatter()
+        
+        if components.year > 0 {
+            return String(format: NSLocalizedString("%@", comment: "Relative date for date older than an year."), NSDateFormatter.localizedStringFromDate(self, dateStyle: NSDateFormatterStyle.ShortStyle, timeStyle: NSDateFormatterStyle.ShortStyle))
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: This shouldn't be a localized string; it's an absolute date, not a relative date, so NSDateFormatter will take care of this for us. Just use the last half of this line on its own.

File: Utils/TimeConstants.swift [hunk 6]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+        }
+
+        if components.month > 0 {
+            if components.month == 1 {
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: You don't need this nesting.

File: Utils/TimeConstants.swift [hunk 7]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+                return String(format: NSLocalizedString("More than a month ago", comment: "Relative date for dates older than a month and less than two months."))
+            }
+            else {
+                return String(format: NSLocalizedString("%@", comment: "Relative date for date older than two months."), NSDateFormatter.localizedStringFromDate(self, dateStyle: NSDateFormatterStyle.ShortStyle, timeStyle: NSDateFormatterStyle.ShortStyle))
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: Again, this isn't a relative date; it's a short-form absolute, so you don't need a localized string.

Whenever you see `NSLocalizedString("%@"`, you can simplify.

File: Utils/TimeConstants.swift [hunk 8]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+            }
+        }
+
+        if components.weekOfYear > 0 {
+            return String(format: NSLocalizedString("More than a week ago", comment: "Relative date for date older than a week and less than a month."))
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: "Description for a date more than a week ago, but less than a month ago."

File: Utils/TimeConstants.swift [hunk 9]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+        }
+
+        if components.day > 0 {
+            if components.day > 1 {
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: Again, you don't need the nesting — just put the sub-condition first.

File: Utils/TimeConstants.swift [hunk 10]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+                return String(format: NSLocalizedString("This week", comment: "Relative date for date in past week."), String(components.day))
+            } else {
+                return String(format: NSLocalizedString("Yesterday", comment: "Relative date for yesterday."))
+            }
+        }
+
+        if components.hour > 0 || components.minute > 0 {
+            formatter.timeStyle = .ShortStyle
+            return String(format: NSLocalizedString("Today at %@", comment: "Relative date for date older than a minute."), formatter.stringFromDate(self))
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: Doesn't this need to be `.localizedStringFromDate`?

File: Utils/TimeConstants.swift [hunk 11]
```diff
@@ -18,6 +18,45 @@ extension NSDate {
...
+        }
+
+        return String(format: NSLocalizedString("Just now", comment: "Relative date for current time."))
```
Comment (rnewman) [inline Utils/TimeConstants.swift]: "Description for a tab that was visited within the last few moments."

**Comment by rnewman** [comment]
Sorry for taking so long to get to this review!

**Comment by rnewman** [comment]
Merged with cleanup: 465c4e4.

**Pull request closed without merging**